Fix incorrect width and height in core/image block

itsTallulah opened the pull request:
## Description

Addresses incorrect `img` tag `width`|`height` values in `core/image` blocks by filtering `render_block`, where it can get the attachment ID and intended size from the classes of the `img` and `figure` tags.
From there it looks for the intended size's w|h values in the attachment's postmeta, and add them to the `img` tag for the return.
Also fixes incorrect srcsets for block images without having to touch them, since it's upstream of them and supplies the values they need.
Returns original safely and silently on failure. Matches narrowly, and can operate on already-filtered blocks as long as they have the required size and ID classes.
Optional switch for `resize` args can be omitted, but it could be nice to give folks the choice. I defer to plat as to how to make this switchable (or if we should just lose it for simplicity and brevity).

## Changelog Description

### Fixed
- Fixed incorrect image sizes and srcsets for `core/image` blocks on frontend requests

## Pre-review checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally or in Codespaces (or has an appropriate fallback).
- [x] This change works and has been tested on a sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change uses a rollout method to ease with deployment (if applicable - especially for large scale actions that require writes).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [x] I've created a changelog description that aligns with the provided examples.

## Pre-deploy checklist

- [ ] VIP staff: Ensure any alerts added/updated conform to internal standards (see internal documentation). 

## Steps to Test

1. Check out PR / use relevant file as generic mu-plugin
2. Examine any page/post with core/image block/s
3. Examine src of any core/image block (other than size:full)
4. Its `width` and `height` values and srcset should all be congruent with its intended size
5. It should also visually look the same size in the frontend and the editor
6. Repeat as required for any/all registered sizes
7. Optional: register some custom sizes and repeat

itsTallulah committed:
Create fix-block-img-tag-sizes.php

addresses incorrect w/h values in core/image block img tags

itsTallulah committed:
Update fix-block-img-tag-sizes.php

remove preamble comment, remove informal test

sjinks commented on fix-block-img-tag-sizes.php:
```diff
@@ -0,0 +1,67 @@
+        // Add the width and height attributes to the img tag
+        $block_content = preg_replace(
+            '/<img ([^>]+)>/',
+            '<img foo="bar" $1 width="' . esc_attr($width) . '" height="' . esc_attr($height) . '">',
```
Is this `foo="bar"` thing really necessary? :-)

sjinks replied:
Also, can the `<img>` tag contain other attributes (say, `alt`)? If so, is it possible to avoid overwriting them?

itsTallulah replied:
Oops, that foo is a testing leftover! Thanks for the spot. And yeah, the img and figure tags can contain other values and classes too.

itsTallulah replied:
See the first commit: it has the now-removed `add_block_bits`, which I was using for as an informal test for extra values in the tags by filtering a bunch of extra stuff into the block markup just upstream.

sjinks commented on fix-block-img-tag-sizes.php:
```diff
@@ -0,0 +1,67 @@
+        $height = $metadata['sizes'][$size_name]['height'];
+        $new_file = $metadata['sizes'][$size_name]['file'];
+
+        if ($resizearg) {
```
This condition is always `false` because we don't seem to set `$resizearg` to `true` anywhere.

itsTallulah replied:
Indeed, we don't set it true anywhere - yet. I'm deferring to y'all as to how we should let the user set this (if we include it at all).

t-wright commented on fix-block-img-tag-sizes.php:
```diff
@@ -0,0 +1,67 @@
+    }
+    return $block_content;
+}
+add_filter('render_block', 'fix_img_block_sizes', 10, 3);
```
Since we're only handling `core/image` blocks, perhaps we can hook into `render_block_core/image` here? https://github.com/WordPress/wordpress-develop/blob/50af37a9083f003f8e98d089091d2cc428797cc5/src/wp-includes/class-wp-block.php#L581

itsTallulah replied:
Good point - Should behave exactly the same on that filter, since it's right after the first one and uses the same values. Will give that a try too.

itsTallulah replied:
Works well and makes the function smaller. Nice, thx!

itsTallulah committed:
remove testing foobar

I'd left a foo=bar in the returned img tag from testing on non-VIP envs, so I took it out.

itsTallulah commented:
Looking at the `resizearg` option I had in here, it would arguably make more sense to have that as a separate filter combo that can hit all contexts where image scaler args would otherwise be `w=` style, so I'm going to remove it from this. I'll pitch that as a separate PR, since it's far lower prio than this fix.

itsTallulah committed:
Update fix-block-img-tag-sizes.php

removed resizearg switch, used render_block_core/image filter, added skip for fullsize, prepended vip to function name to avoid collisions with existing fix_img_block_sizes filters in use

itsTallulah commented:
Removed `resizearg` section, added a skip for fullsize image size, changed filter to `render_block_core/image`, renamed function to `vip_fix_img_block_sizes` for specificity and to avoid collisions (since pervious versions of this fix have been shipped to some customers as a stopgap).

itsTallulah commented:
Bumping out of stale to discuss

itsTallulah committed:
switch to HTML API

itsTallulah commented:
Refactored to use HTML API as discussed in Slack. Verbose with comments on purpose, since I haven't used the HTML API much and I wanted to make sure the flow was followable. Thoughts?

rinatkhaziev committed:
Fix lint issues

rinatkhaziev committed:
Move the functionality inside a8c-files.php

rinatkhaziev committed:
Add tests

rinatkhaziev committed:
Add dataProvider

rinatkhaziev committed:
Fix tests

rinatkhaziev committed:
Cleanup

rinatkhaziev committed:
Fix admin-context tests

sjinks commented on a8c-files.php:
```diff
@@ -104,6 +104,9 @@ public function __construct() {
 		}
+
+		// Add filter to fix image block sizes
+		add_filter( 'render_block_core/image', array( $this, 'fix_img_block_sizes' ), 10, 3 );
```
```suggestion
		if ( ! is_admin() ) {
			add_filter( 'render_block_core/image', array( $this, 'fix_img_block_sizes' ), 10, 3 );
		}
```

and remove `is_admin()` check in `fix_img_block_sizes()`.

sjinks commented on a8c-files.php:
```diff
@@ -511,6 +514,88 @@ public function update_attachment_meta() {
+	 * @return string Modified block content.
+	 */
+	public function fix_img_block_sizes( $block_content, $block, $instance ) {
+		// Don't fire in wp-admin, image blocks are fine there
+		if ( is_admin() ) {
+			return $block_content;
+		}
+
```
```suggestion
```

sjinks commented on a8c-files.php:
```diff
@@ -511,6 +514,88 @@ public function update_attachment_meta() {
+
+		// Set width/height for img
+		if ( $update_tag_processor->next_tag( 'img' ) ) {
+			$update_tag_processor->set_attribute( 'width', esc_attr( $width ) );
+			$update_tag_processor->set_attribute( 'height', esc_attr( $height ) );
```
```suggestion
			$update_tag_processor->set_attribute( 'width', $width );
			$update_tag_processor->set_attribute( 'height', $height );
```

See https://developer.wordpress.org/reference/classes/wp_html_tag_processor/set_attribute/:
> For string attributes, the value is escaped using the `esc_attr` function.

See https://github.com/WordPress/wordpress-develop/blob/50fb4086b7afbfa012c5d1f2eeff79b1bae3b00e/src/wp-includes/html-api/class-wp-html-tag-processor.php#L3875

sjinks commented on a8c-files.php:
```diff
@@ -511,6 +514,88 @@ public function update_attachment_meta() {
+		// Get values ready
+		$width                = $metadata['sizes'][ $size_name ]['width'];
+		$height               = $metadata['sizes'][ $size_name ]['height'];
+		$update_tag_processor = new WP_HTML_Tag_Processor( $block_content );
+
+		// Set width/height for img
+		if ( $update_tag_processor->next_tag( 'img' ) ) {
+			$update_tag_processor->set_attribute( 'width', esc_attr( $width ) );
+			$update_tag_processor->set_attribute( 'height', esc_attr( $height ) );
+		}
+
+		// Get the result ready
+		$block_content = $update_tag_processor->get_updated_html();
```
Do you think we can reuse `$img_tag_processor` here?

sjinks commented on tests/files/test-fix-block-img-tag-sizes.php:
```diff
@@ -0,0 +1,110 @@
+		if ( $is_admin && ! defined( 'WP_ADMIN' ) ) {
+			define( 'WP_ADMIN', true );
+		}
+
```
I don't like it because it might affect other tests.

I suggest moving the `is_admin()` check around the `add_filter()` call.

sjinks commented on tests/files/test-fix-block-img-tag-sizes.php:
```diff
@@ -0,0 +1,110 @@
+		$actual = $files->fix_img_block_sizes( $block_content, array(), null );
+
+		// For the successful-update case, add more detailed assertions
+		if ( $metadata ) {
+			$this->assertStringContainsString( 'width="300"', $actual, 'Width attribute should be present' );
+			$this->assertStringContainsString( 'height="200"', $actual, 'Height attribute should be present' );
+			$this->assertStringContainsString( 'class="wp-image-123"', $actual, 'Image class should be preserved' );
+			$this->assertStringContainsString( 'class="wp-block-image size-medium"', $actual, 'Figure class should be preserved' );
+			remove_filter( 'get_post_metadata', $filter, 10 );
+		} else {
+			$this->assertEquals( $expected, $actual );
+		}
```
Maybe we can make `$expected` a `string` or an `array`? If `$expected` is a string, we call `assertEquals()`. Otherwise, we invoke `assertStringContainsString()` for every array item. This will make the code cleaner.

rinatkhaziev replied:
@sjinks there's no guarantee order of arguments will stay in the same order (hence the THIS IS FLAKY comment). With this approach it's less flaky

sjinks replied:
@rinatkhaziev we will have `[ 'width="300"', 'height="200"', ... ]` etc and then

```php
if ( is_array(‌ $expected ) ) {
    foreach ( $expected as $substring ) {
        self::assertStringContainsString( $substring, $actual );
    }
}
```

This will move all hardcoded checks out of the test into the data.

sjinks replied:
https://github.com/Automattic/vip-go-mu-plugins/blob/4ddac9fb0ad018436c51be39d1617f5806671ac6/tests/files/test-fix-block-img-tag-sizes.php#L91-L97

sjinks commented on tests/files/test-fix-block-img-tag-sizes.php:
```diff
@@ -0,0 +1,110 @@
+		}
+
+		// Mock wp_get_attachment_metadata if metadata is provided
+		if ( $metadata ) {
+			$attachment_id = 123; // Matches the wp-image-123 in test data
+			$filter        = function ( $meta_value, $post_id, $meta_key ) use ( $metadata, $attachment_id ) {
+				if ( '_wp_attachment_metadata' === $meta_key && $attachment_id == $post_id ) {
+					return [ $metadata ];
+				}
+				return $meta_value;
+			};
+			add_filter( 'get_post_metadata', $filter, 10, 3 );
+		}
```
Should this go to `setUp()` and `tearDown()`?

ThirdEyeRose commented:
Hi folks - we have a customer who is interested in using this solution when we release it. I see we've moved into a reviewing stage but I know we're still in active development. Is there a rough estimate for when this is expected to be in the pipeline to production?

The customer has been trying to adapt this concept to their own custom plugin, but if our version will be available in a few weeks, they'd prefer to wait for our version, so I wanted to check in.

rinatkhaziev committed:
Update a8c-files.php

Co-authored-by: Volodymyr Kolesnykov <volodymyr.kolesnykov@automattic.com>

sjinks committed:
fix: address my own suggestions

sjinks committed:
test: refactor

sjinks committed:
test: do not run tests in a separate process
